http-client/Network/HTTP/Client/Connection.hs

```diff
@@ -184,10 +189,31 @@ openSocket tweakSocket addr =
             NS.connect sock (NS.addrAddress addr)
             return sock)
 
+-- Pick up an IP using an approximation of the happy-eyeballs algorithm:
+-- https://datatracker.ietf.org/doc/html/rfc8305
+-- 
 firstSuccessful :: [NS.AddrInfo] -> (NS.AddrInfo -> IO a) -> IO a
-firstSuccessful []     _  = error "getAddrInfo returned empty list"
-firstSuccessful (a:as) cb =
-    cb a `E.catch` \(e :: E.IOException) ->
-        case as of
-            [] -> E.throwIO e
-            _  -> firstSuccessful as cb
+firstSuccessful []        _  = error "getAddrInfo returned empty list"
+firstSuccessful addresses cb = do
+    result <- newEmptyMVar
+    either E.throwIO pure =<<
+        withAsync (tryAddresses result)
+            (\_ -> takeMVar result)
+  where
+    -- https://datatracker.ietf.org/doc/html/rfc8305#section-5
+    connectionAttemptDelay = 250 * 1000
+
+    tryAddresses result = do
+        lastException <- take 1 . filter isLeft <$> go addresses []
+        for_ lastException $ tryPutMVar result
+      where
+        go []             asyncs = mapM wait asyncs
```

Comment:
This logic doesn't look right to me. It seems like it's going to guarantee that all `addrs` get a forked thread.

Reply:
Yes, pretty much. The only way to prevent it is to fill up the `result` then `go` gets interrupted. Another option would be to do something like
```
z <- isEmptyMVar result
if z 
   then go addrs $ a : asyncs
   else pure []
```
but it clutters the code and I'm not sure if it makes sense.